tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc

```diff
@@ -0,0 +1,216 @@
+      TF_RETURN_IF_ERROR(b->AddInputDataset(ctx, input_, &input_node));
+
+      std::vector<Node*> other_arguments;
+      other_arguments.reserve(captured_func_->captured_inputs().size());
+      DataTypeVector other_arguments_types;
+      other_arguments_types.reserve(captured_func_->captured_inputs().size());
+      for (const Tensor& t : captured_func_->captured_inputs()) {
+        Node* node;
+        DatasetBase* input;
+        Status s = GetDatasetFromVariantTensor(t, &input);
+        if (s.ok()) {
+          TF_RETURN_IF_ERROR(b->AddInputDataset(ctx, input, &node));
+        } else {
+          TF_RETURN_IF_ERROR(b->AddTensor(t, &node));
+        }
+        other_arguments.emplace_back(node);
+        other_arguments_types.emplace_back(t.dtype());
+      }
+      AttrValue f_attr;
+      b->BuildAttrValue(func_, &f_attr);
+
+      AttrValue other_arguments_types_attr;
+      b->BuildAttrValue(other_arguments_types, &other_arguments_types_attr);
+
+      TF_RETURN_IF_ERROR(b->AddDataset(
+            this, {std::make_pair(0, input_node)},
+            {std::make_pair(1, other_arguments)},
+            {std::make_pair("predicate", f_attr),
+             std::make_pair("Targuments", other_arguments_types_attr)},
+             output));
+      return Status::OK();
+    }
+
+   private:
+    class Iterator : public DatasetIterator<Dataset> {
+     public:
+      explicit Iterator(const Params& params)
+          : DatasetIterator<Dataset>(params) {}
+
+      Status Initialize(IteratorContext* ctx) override {
+        TF_RETURN_IF_ERROR(
+            dataset()->input_->MakeIterator(ctx, prefix(), &input_impl_));
+        return dataset()->captured_func_->Instantiate(
+            ctx, &instantiated_captured_func_);
+      }
+
+      Status GetNextInternal(IteratorContext* ctx,
+                             std::vector<Tensor>* out_tensors,
+                             bool* end_of_sequence) override {
+        {
+          tf_shared_lock l(mu_);
+          if(!input_impl_) {
+            *end_of_sequence = true;
+            return Status::OK();
+          }
+          TF_RETURN_IF_ERROR(
+              input_impl_->GetNext(ctx, out_tensors, end_of_sequence));
+        }
+        if (*end_of_sequence) {
+          mutex_lock l(mu_);
+          input_impl_.reset();
+          return Status::OK();
+        }
+
+        std::vector<Tensor> bool_output;
+
+        Status s = instantiated_captured_func_->RunWithBorrowedArgs(
+            ctx, *out_tensors, &bool_output);
+
+        if (s.ok()) {
+          if(bool_output.size() != 1 || bool_output[0].dtype() != DT_BOOL ||
+             bool_output[0].NumElements() != 1) {
+            return errors::InvalidArgument(
+                "`predicate` must returns a scalar bool tensor.");
+          }
+          if (!bool_output[0].scalar<bool>()()) {
+            *end_of_sequence = true;
+            return Status::OK();
+          }
+        }
+        return s; // propagate error to caller
```

Comment:
nit: if `predicate(out_tensors)` returns `true`, the control-flow will reach here which means the comment "propage error to caller" is technically inaccurate because `s` could be `Status::OK()`.

Consider modifying lines 162 through 165 as follows:

```
*end_of_sequence = !bool_output[0].scalar<bool>()();
return Status::OK();
```

This will a) always explicitly set `end_of_sequence` and b) avoid the potentially misleading comment on line 167